community/server/src/test/java/org/neo4j/server/NeoServerRestartTest.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.server;
+
+import org.junit.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import org.neo4j.io.pagecache.PageEvictionCallback;
+import org.neo4j.io.pagecache.PageSwapper;
+import org.neo4j.io.pagecache.impl.SingleFilePageSwapperFactory;
+import org.neo4j.test.ThreadTestUtils;
+
+import static org.junit.Assert.fail;
+
+public abstract class NeoServerRestartTest
+{
+
+    public static final String CUSTOM_SWAPPER = "CustomSwapper";
+    private static volatile Semaphore semaphore;
+
+    static
+    {
+        semaphore = new Semaphore( 0 );
+    }
+
+    @Test
+    public void shouldBeAbleToRestartWhenStoppedDuringStartup() throws IOException, InterruptedException
+    {
+        semaphore.drainPermits();
+        NeoServer server = getNeoServer( CUSTOM_SWAPPER );
+
+        CountDownLatch latch = new CountDownLatch( 1 );
+        ThreadTestUtils.fork( stopServerAfterStartingHasStarted( server, latch ) );
+        server.start();
+        //Wait for the server to stop.
+        if ( !latch.await( 20, TimeUnit.SECONDS ) )
+        {
+            fail( "Timed out while waiting for server to stop" );
+        }
+
+        server = getNeoServer( CUSTOM_SWAPPER );
+        server.start();
+        server.stop();
+    }
+
+    protected abstract NeoServer getNeoServer( String customPageSwapperName ) throws IOException;
+
+    private Runnable stopServerAfterStartingHasStarted( NeoServer server, CountDownLatch latch )
+    {
+        return () ->
+        {
+            try
+            {
+                semaphore.acquire();
+                server.stop();
```

Comment:
if your stop will throw any exception you will not countDown your latch.
Empty catch block claims that any kind of exception is fine - in that case we should not fail test on latch awaiting timeout.